File: big_data_learning/spark-learning/com/shawn/study/spark/core/api/map-partitions-demo.py
from pyspark import SparkContext, SparkConf


def get_max(iterator):
    return [max(iterator)]


if __name__ == '__main__':
    conf = SparkConf().setAppName(value="map-partition-demo").setMaster(value="local[*]")
    context = SparkContext(conf=conf)
    rdd = context.parallelize(c=[1, 2, 3, 4, 5, 6, 7, 8], numSlices=2)
    print(rdd.mapPartitions(f=get_max).collect())
    context.stop()

File: big_data_learning/spark-scala-java-learning/src/main/java/com/shawn/study/big/data/spark/java/core/rdd/basic/MapPartitionsDemo.java
package com.shawn.study.big.data.spark.java.core.rdd.basic;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import org.apache.commons.lang.StringUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

public class MapPartitionsDemo {

  public static void main(String[] args) {
    SparkConf conf = new SparkConf().setAppName("map-partition-demo").setMaster("local[*]");
    JavaSparkContext context = new JavaSparkContext(conf);
    JavaRDD<Integer> javaRDD = context.parallelize(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8), 2);
    List<Integer> list =
        javaRDD
            .mapPartitions(
                iterator -> {
                  int max = 0;
                  while (iterator.hasNext()) {
                    Integer next = iterator.next();
                    max = Math.max(max, next);
                  }
                  return Collections.singletonList(max).iterator();
                })
            .collect();
    System.out.println(StringUtils.join(list, ", "));
    context.close();
  }
}
